### PR TITLE
Implement Complete Signup Functionality with Comprehensive Tests

### PR DESCRIPTION
**Pull request by Nagalakshmi136**
### Overview

This PR delivers a full-featured signup flow for our authentication module. It covers the entire process—from collecting user data in the UI, performing thorough form validation, making API calls, handling errors gracefully, and managing state using Riverpod, to comprehensive unit and widget tests. Additionally, this PR refines our token refresh logic in the AuthInterceptor, ensuring robust error handling and reliable navigation for expired tokens.

### Key Features

- **Complete Signup Flow:**  
  - Implements user signup from form submission through backend account creation and email verification.
  - Integrates with our backend API using Dio via an `AuthRepository` that leverages a common `ApiCallHandler` for consistent error handling.
  - Fields in sign up page:
    *   Username
    *   Email (using `Validators.email`)
    *   Password (using `Validators.password`)
    *   Confirm Password (using `Validators.confirmPassword`)
    *   Phone Number (using `Validators.phoneNumber`)
    *   Date of Birth (with date picker)
    *   Gender (using `CustomDropdown`)

- **Form Validation:**  
  - Uses a combination of custom widgets (e.g., `CustomTextField`, `CustomDropDownBox`) and validators (`Validators.email`, `Validators.password`, etc.) to enforce proper input for username, email, password, phone number, date of birth, and gender.
  - Provides immediate feedback to the user for invalid entries.

- **Riverpod State Management:**  
  - Utilizes `AuthNotifier` (a `StateNotifier`) and `AuthState` to manage and expose the current authentication status (loading, authenticated, error, verification pending, etc.).
  - Ensures that state transitions trigger the appropriate UI changes (e.g., navigation to verification or home screens).

- **Robust Error Handling:**  
  - Maps low-level exceptions (e.g., DioExceptions) to domain-specific exceptions (e.g., `SignupFailedException`, `EmailNotVerifiedException`).
  - Centralizes API error handling via `ApiCallHandler` and `_executeAuthOperation`, reducing code duplication.
  - Uses logging (via Logger) for improved traceability during signup and token refresh processes.

- **Token Refresh & AuthInterceptor Enhancements:**  
  - Automatically adds the access token to outgoing requests and, upon receiving a 401 error, attempts a token refresh.
  - Queues failing requests while a refresh is in progress and retries them once a new token is obtained.
  - If token refresh fails (e.g., due to an expired refresh token), clears authentication data and logs the user out.

- **Comprehensive Testing:**  
  - **Unit Tests:**  
    - For `AuthNotifier`: Covering success, error, and edge cases for signup, sending verification codes, and resetting passwords.
    - For `AuthRepository`: Ensuring correct API calls and error mapping.
    - For `AuthInterceptor`: Simulating token refresh scenarios, queued request retries, and refresh failures.
    - For `Validators`: Ensures validations for inputs in the signup page.
  - Mocks and fakes are set up using Mocktail, ensuring isolated, deterministic tests.

## Output
Screenshot from 2025-01-31 01-29-00
Screenshot from 2025-01-31 01-29-11
Screenshot from 2025-01-31 01-31-06
Screenshot from 2025-01-31 01-29-22

## Summary by CodeRabbit

- **New Features**
  - Rebranded the app to "Paper Trading App" with refreshed texts for login, registration, and reset password flows.
  - Introduced new authentication screens including initial, verification, and home pages.
  - Added new UI components such as a customizable primary button, a loading indicator, and enhanced snackbar notifications.
  - Implemented a structured approach for managing authentication state and operations with comprehensive logging and error handling.
  - Introduced a new `SignupRequest` model for user registration and a `TokenStorageService` for managing authentication tokens.
  - Added a new `ApiEndpoints` class for organized API endpoint management.

- **Improvements**
  - Updated welcome page layout with better spacing, responsive design, and improved navigation.
  - Enhanced form validations and overall UI responsiveness with dark theme support and optimized keyboard dismissal.
  - Improved error handling and user feedback mechanisms throughout the authentication processes.
  - Enhanced API error handling with a dedicated `ApiErrorHandler` class for better user experience.
  - Streamlined text styling across the application with the introduction of a flexible `customTextStyle` method.

### CODE REVIEW

File: frontend/lib/features/auth/application/model/signup_request.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  SignupRequest({
+    required this.email,
+    required this.username,
+    required this.password,
+    required this.confirmPassword,
+    required this.dateOfBirth,
+    required this.phoneNumber,
+    required this.gender,
+  });
+
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/application/model/signup_request.dart:35]: i added the validations in a seperate file

File: frontend/lib/features/auth/application/providers/auth_providers.dart [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+final authRepositoryProvider = Provider((ref) {
+  final dio = ref.read(dioProvider);
+  final secureStorage = ref.read(secureStorageProvider);
+  final apiCallHandler = ApiCallHandler();
+  return AuthRepository(
+    dio: dio,
+    tokenStorage: secureStorage,
+    apiCallHandler: apiCallHandler,
+  );
+});
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/application/providers/auth_providers.dart]: changed

File: frontend/lib/features/home/home.dart [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+final protectedDataProvider = FutureProvider<Map<String, dynamic>>((ref) async {
+  final dio = ref.read(dioProvider); // Get Dio instance with auth interceptor
+  try {
+    final response =
+        await dio.get('/authentication/protected-endpoint'); // API Call
+    return response.data;
+  } catch (e) {
+    throw Exception("Failed to fetch protected data");
+  }
+});
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/home/home.dart:14]: its a basic testing page, no changes needed.

File: frontend/lib/core/utils/validators.dart [hunk 4]
```diff
@@ -32,4 +35,25 @@ class Validators {
...
+  static String? phoneNumber(String? value) {
+    if (value == null || value.isEmpty) {
+      return 'Please enter your phone number';
+    }
+    final phoneRegExp = RegExp(r'^\+?\d{10}$');
+    if (!phoneRegExp.hasMatch(value)) {
+      return 'Please enter a valid phone number';
+    }
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/utils/validators.dart]: currently not needed.

File: frontend/lib/core/utils/handlers/api_call_handler.dart [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    } catch (e, stackTrace) {
+      _logger.e('Unexpected error during $operationName: $e', e, stackTrace);
+      if (!suppressError) {
+        throw exception('An unexpected error occurred');
+      } else {
+        return null as T; // Explicitly handle null case
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/utils/handlers/api_call_handler.dart]: No need of it. I seperately used an error handler.

File: frontend/test/core/network/dio_client_test.dart [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    test('PrettyDioLogger is added in debug mode', () {
+      // Use runZoned to override kDebugMode *within* the test
+      return runZoned(() {
+        dioClient = DioClient(mockSecureStorage, mockRef);
+        final dio = dioClient.dio;
+        expect(dio.interceptors, isNotEmpty);
+        expect(dio.interceptors.any((i) => i is PrettyDioLogger), isTrue);
+        // Force debug
+      });
+    });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/core/network/dio_client_test.dart]: added

File: frontend/test/utils/hadler/api_error_handler_test.dart [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        // Act
+        final result = ApiErrorHandler.handleInvalidResponse(mockResponse);
+
+        // Assert
+        expect(
+            result,
+            isA<AppException>()
+                .having((e) => e.message, 'message', 'Invalid response (404)'));
+      });
+    });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/utils/hadler/api_error_handler_test.dart]: added

File: frontend/test/features/auth/functionality/repository/auth_repository_test.dart [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+            throwsA(isA<LogoutFailedException>()),
+          );
+
+          verify(() => mockApiCallHandler.handleApiCall<void>(
+                call: any(named: 'call'),
+                exception: any(named: 'exception'),
+                operationName: any(named: 'operationName'),
+              )).called(1);
+          verify(() => mockSecureStorageService.clearTokens()).called(1);
+        });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/features/auth/functionality/repository/auth_repository_test.dart]: changed

File: frontend/lib/features/auth/presentation/pages/verify_account.dart [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+      // DON'T show the success snackbar here.  We'll handle it in ref.listen
+      _isCodeSent = true; // Mark the code as sent.
+    } else if (_email.isEmpty) {
+      // Use addPostFrameCallback to ensure Navigator.pop() is called after build
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        Navigator.of(context).pop(); // Go back
+        context.showErrorSnackBar("Error: Email address is missing.");
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/presentation/pages/verify_account.dart]: added

File: frontend/lib/features/auth/presentation/pages/verify_account.dart [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+      onPressed: _onResendOtpPressed,
+      child: Text(
+        "Resend OTP",
+        style: theme.textTheme.labelLarge!.copyWith(
+          // Use theme for styling
+          color: theme.colorScheme.primary,
+        ),
+      ),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/presentation/pages/verify_account.dart]: i think now its enough, backend can verify it.

File: frontend/lib/shared/buttons/primary_button.dart [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    this.borderRadius = const BorderRadius.all(Radius.circular(10)),
+    this.child,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+    return GestureDetector(
+      onTap: onPressed, // Use onTap directly, handles null for disabled state
+      child: Container(
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/shared/buttons/primary_button.dart:59]: gesture detected give me more flexibility for appearance and functionlity

File: frontend/test/utils/validators_test.dart [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        () {
+      final result = Validators.phoneNumber('+11234567890');
+      expect(result, "Please enter a valid phone number");
+    });
+
+    test('should return null when phone number is valid', () {
+      final result = Validators.phoneNumber('1234567890');
+      expect(result, null);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/utils/validators_test.dart:180]: now its enough.

File: frontend/lib/features/auth/presentation/pages/register_page.dart [hunk 13]
```diff
@@ -41,121 +59,207 @@ class _RegisterPageState extends State<RegisterPage> {
...
     );
     if (pickedDate != null) {
-      setState(() {
-        _dateController.text =
-            "${pickedDate.day}/${pickedDate.month}/${pickedDate.year}";
-      });
+      // Format the date as needed.  Using a consistent format is important.
+      _controllers['date']!.text =
+          "${pickedDate.day}/${pickedDate.month}/${pickedDate.year}";
     }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/presentation/pages/register_page.dart:93]: backend accepts this format

File: frontend/pubspec.yaml [hunk 14]
```diff
@@ -34,8 +34,16 @@ dependencies:
 
   # The following adds the Cupertino Icons font to your application.
   # Use with the CupertinoIcons class for iOS style icons.
-  cupertino_icons: ^1.0.8
-
+  dio: ^5.7.0
+  flutter_secure_storage: ^9.2.4
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/pubspec.yaml:38]: added

File: frontend/lib/features/auth/presentation/pages/register_page.dart [hunk 15]
```diff
@@ -41,121 +59,207 @@
...
+        password: _controllers['password']!.text,
+        confirmPassword: _controllers['confirmPassword']!.text,
+        phoneNumber: _controllers['phone']!.text,
+        dateOfBirth: _controllers['date']!.text,
+        gender: _selectedGender!,
+      );
+
+      await ref.read(authNotifierProvider.notifier).signup(request);
     }
   }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/features/auth/presentation/pages/register_page.dart]: already i added these validations.

File: frontend/lib/core/constants/api_endpoints.dart [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+  static const sendVerification = '/authentication/send-verification-code';
+  static const verifyCode = '/authentication/verify-code';
+  static const refreshToken = '/authentication/refresh-token';
+  static const signin = '/authentication/signin';
+  static const logout = '/authentication/logout';
+  static const signup = '/authentication/signup';
+  static const protected = '/authentication/protected-endpoint';
+  static const sendResetPasswordCode =
+      '/authentication/send-reset-password-code';
+  static const resetPassword = '/authentication/reset-password';
```
Comment (Munikumar09) [inline frontend/lib/core/constants/api_endpoints.dart]: It seems `authentication` is used in many times which causes typos. Is it possible to assign a constant and use it like 
```
static const router = '/authentication'
static const sendVerification = '$router/send-verification-code'
```
It also keeps the DRY principle

File: frontend/lib/core/constants/api_endpoints.dart [hunk 17]
```diff
@@ -0,0 +1,15 @@
+abstract class ApiEndpoints {
+  static const String baseUrl = "http://10.0.2.2:8000";
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/constants/api_endpoints.dart]: present its enough

File: frontend/lib/core/constants/api_endpoints.dart (hunk 16 above)
Comment (Nagalakshmi136) [reply frontend/lib/core/constants/api_endpoints.dart]: added

File: frontend/lib/core/constants/api_endpoints.dart (hunk 17 above)
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/constants/api_endpoints.dart:18]: its just in local url

File: frontend/lib/core/utils/handlers/api_call_handler.dart [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+      }
+    } catch (e, stackTrace) {
+      _logger.e('Unexpected error during $operationName: $e', e, stackTrace);
+      if (!suppressError) {
+        throw exception('An unexpected error occurred');
+      } else {
+        return null as T;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/utils/handlers/api_call_handler.dart]: changed

File: frontend/lib/core/network/interceptor/auth_interceptor.dart [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+  static const refreshTokenEndpoint = ApiEndpoints.refreshToken;
+
+  final SecureStorageService _secureStorage;
+  final Dio _dio;
+  final Ref ref;
+  final Logger _logger = Logger();
+
+  bool _isRefreshing = false;
+  Completer<String>? _refreshCompleter;
+  final List<_QueuedRequest> _requestQueue = [];
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/network/interceptor/auth_interceptor.dart:64]: added

File: frontend/lib/core/network/interceptor/auth_interceptor.dart [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+            response: response,
+            type: DioExceptionType.badResponse,
+            error: errorMessage);
+      }
+    } on DioException catch (e) {
+      _logger.e("_refreshAccessToken: DioException: ${e.message}", e);
+      await _clearAuthDataAndLogout(e);
+      rethrow;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/lib/core/network/interceptor/auth_interceptor.dart]: added

File: frontend/test/features/auth/functionality/repository/auth_repository_test.dart [hunk 21]
```diff
@@ -0,0 +1,977 @@
...
+            final exceptionFunction =
+                invocation.namedArguments[const Symbol('exception')]
+                    as AppException Function(String);
+            final String errorMessage =
+                e.response?.data['message']?.toString() ??
+                    "Sending reset password code failed";
+            capturedException = exceptionFunction(errorMessage);
+            throw capturedException!;
+          }
+        });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/features/auth/functionality/repository/auth_repository_test.dart]: added

File: frontend/test/features/auth/functionality/repository/auth_repository_test.dart [hunk 22]
```diff
@@ -0,0 +1,977 @@
...
+          //   try {
+          //     return await invocation.namedArguments[const Symbol('call')]();
+          //   } on DioException catch (e) {
+          //     final exceptionFunction =
+          //         invocation.namedArguments[const Symbol('exception')]
+          //             as AppException Function(String);
+          //     final String errorMessage = e.error.toString();
+          //     throw exceptionFunction(errorMessage);
+          //   }
+          // });
```
[the remark this answers is not in the document]
Comment (Nagalakshmi136) [reply frontend/test/features/auth/functionality/repository/auth_repository_test.dart]: removed

File: frontend/lib/core/constants/app_text_styles.dart [hunk 23]
```diff
@@ -1,6 +1,32 @@
...
+  • bodyText2(color):
+      - Returns a TextStyle for secondary body text.
+  • caption(color):
+      - Returns a TextStyle for caption texts.
+*/
+
+// Code:
 import 'package:flutter/material.dart';
 
+/// Provides helper methods that return TextStyle objects for different use cases in the application.
```
Comment (Munikumar09) [inline frontend/lib/core/constants/app_text_styles.dart]: Why can't we use single method for all the headings by passing the parameters? Something like the below
```
/// Can be placed in some const file
static const heading1FontSize = 30
static const heading1FontWeight  = FontWeight.w700
class AppTextStyles {
  static TextStyle headline(Color color, fontSize: int, fontWeight: int) {
    return TextStyle(
      color: color,
      fontSize: fontSize,
      fontWeight: FontWeight,
    );
  }
  ```

File: frontend/lib/core/network/interceptor/auth_interceptor.dart [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+  AuthInterceptor(this._secureStorage, this._dio, this.ref);
+
+  /// onRequest attaches the Authorization header if an access token exists.
+  @override
+  Future<void> onRequest(
+      RequestOptions options, RequestInterceptorHandler handler) async {
+    final tokens = await _secureStorage.getTokens();
+    final accessToken = tokens.accessToken;
+    if (accessToken != null && !options.path.contains(refreshTokenEndpoint)) {
+      options.headers[_authorizationHeader] = _bearerPrefix + accessToken;
```
Comment (Munikumar09) [inline frontend/lib/core/network/interceptor/auth_interceptor.dart]: why do we need to create a variable `accessToken` as it is already available in `tokens` object?

Comment (muni7085) [reply frontend/lib/core/network/interceptor/auth_interceptor.dart]: removed

File: frontend/lib/core/constants/app_text_styles.dart (hunk 23 above)
Comment (muni7085) [reply frontend/lib/core/constants/app_text_styles.dart]: changed
